File: backend/src/utils/geo.ts
const EARTH_RADIUS_M = 6371000;

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180;
}

// Haversine 거리(m). MVP 규모(수십 곳)에서는 PostGIS 없이 애플리케이션 레벨 계산으로 충분하다.
export function haversineDistanceM(a: { lat: number; lng: number }, b: { lat: number; lng: number }): number {
  const dLat = toRadians(b.lat - a.lat);
  const dLng = toRadians(b.lng - a.lng);
  const lat1 = toRadians(a.lat);
  const lat2 = toRadians(b.lat);

  const h = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.sqrt(h));
}

export function isBakeryOpenNow(openingHours: string | null): boolean {
  if (!openingHours) return false;
  const match = openingHours.match(/(\d{2}):(\d{2})-(\d{2}):(\d{2})/);
  if (!match) return false;

  const [, startH, startM, endH, endM] = match.map(Number) as unknown as [never, number, number, number, number];
  const now = new Date();
  const minutesNow = now.getHours() * 60 + now.getMinutes();
  const startMinutes = startH * 60 + startM;
  const endMinutes = endH * 60 + endM;

  return minutesNow >= startMinutes && minutesNow <= endMinutes;
}

File: ppangkal-frontend/lib/core/geo.dart
import 'dart:math' as math;

/// Great-circle distance in meters (same formula as the backend's
/// `utils/geo.ts` `haversineDistanceM`).
double haversineM(double lat1, double lng1, double lat2, double lng2) {
  const earthRadiusM = 6371000.0;
  double rad(double deg) => deg * math.pi / 180;
  final dLat = rad(lat2 - lat1);
  final dLng = rad(lng2 - lng1);
  final a = math.pow(math.sin(dLat / 2), 2) +
      math.cos(rad(lat1)) * math.cos(rad(lat2)) * math.pow(math.sin(dLng / 2), 2);
  return 2 * earthRadiusM * math.asin(math.sqrt(a));
}
